src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs

```diff
@@ -235,35 +250,32 @@ public override string ToString()
 
             var vsb = new ValueStringBuilder(stackalloc char[512]);
 
-            if (_headerStore is Dictionary<HeaderDescriptor, object> headerStore)
+            foreach (HeaderEntry entry in GetEntries())
             {
-                foreach (KeyValuePair<HeaderDescriptor, object> header in headerStore)
-                {
-                    vsb.Append(header.Key.Name);
-                    vsb.Append(": ");
+                vsb.Append(entry.Key.Name);
+                vsb.Append(": ");
 
-                    GetStoreValuesAsStringOrStringArray(header.Key, header.Value, out string? singleValue, out string[]? multiValue);
-                    Debug.Assert(singleValue is not null ^ multiValue is not null);
+                GetStoreValuesAsStringOrStringArray(entry.Key, entry.Value, out string? singleValue, out string[]? multiValue);
+                Debug.Assert(singleValue is not null ^ multiValue is not null);
 
-                    if (singleValue is not null)
-                    {
-                        vsb.Append(singleValue);
-                    }
-                    else
-                    {
-                        // Note that if we get multiple values for a header that doesn't support multiple values, we'll
-                        // just separate the values using a comma (default separator).
-                        string? separator = header.Key.Parser is HttpHeaderParser parser && parser.SupportsMultipleValues ? parser.Separator : HttpHeaderParser.DefaultSeparator;
+                if (singleValue is not null)
+                {
+                    vsb.Append(singleValue);
+                }
+                else
+                {
+                    // Note that if we get multiple values for a header that doesn't support multiple values, we'll
+                    // just separate the values using a comma (default separator).
+                    string? separator = entry.Key.Parser is HttpHeaderParser parser && parser.SupportsMultipleValues ? parser.Separator : HttpHeaderParser.DefaultSeparator;
 
-                        for (int i = 0; i < multiValue!.Length; i++)
-                        {
-                            if (i != 0) vsb.Append(separator);
-                            vsb.Append(multiValue[i]);
-                        }
+                    for (int i = 0; i < multiValue!.Length; i++)
+                    {
+                        if (i != 0) vsb.Append(separator);
+                        vsb.Append(multiValue[i]);
                     }
```

Comment:
`GetStoreValuesAsStringOrStringArray` does special-case `length == 0` to avoid allocating an empty array, but I don't think it's reachable. The entry should never be empty - it should have been removed when the last value is removed. I'll double-check the code and add asserts.

https://github.com/dotnet/runtime/blob/bbbfba31b61a2299c434172f709716325c073ef8/src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs#L1131